ProtocolDataUnits/pdu.py

```diff
@@ -1,251 +1,618 @@
-    :param payload: The actual data carried by the PDU.
-
-    :nested_pdu: PDU
-    :param nested_pdu: An optional nested PDU object.
-
-    :byte_order: str
-    :param byte_order: The byte order for encoding/decoding operations.
-
-    :metadata: dict
-    :param metadata: Additional metadata or information about the PDU.
-    """
-
-    """
-    Represents a Protocol Data Unit (PDU).
+    The PDU (Protocol Data Unit) class represents a data structure that is used for encoding and decoding binary data.
+    It provides methods for defining the structure of the PDU, encoding data into binary format, and decoding binary data
+    back into the original structure.
 
     Attributes:
-    - type (int): The type of the PDU.
-    - duration (int): The duration of the PDU.
-    - payload (bytes): The payload of the PDU.
-    - nested_pdu (PDU): A nested PDU object.
-    - byte_order (str): The byte order for encoding/decoding.
-    - metadata (dict): Additional metadata for the PDU.
+        fields (list): A list of tuples representing the fields of the PDU. Each tuple contains the field type, name,
+                       and optional value.
+        byte_order (str): The byte order used for encoding and decoding the PDU. It can be either 'big' or 'little'.
+        pdu_length (int): The length of the PDU in bytes.
+        default_values (dict): A dictionary containing default values for fields that are not provided during encoding.
+
+    Methods:
+        length(length): Sets the length of the PDU.
+        order(byte_order): Sets the byte order of the PDU.
+        uint8(name, value, default): Adds a uint8 field to the PDU.
+        uint16(name, value, default): Adds a uint16 field to the PDU.
+        uint32(name, value, default): Adds a uint32 field to the PDU.
+        float(name, value, default): Adds a float field to the PDU.
+        double(name, value, default): Adds a double field to the PDU.
+        int8(name, value, default): Adds an int8 field to the PDU.
+        int16(name, value, default): Adds an int16 field to the PDU.
+        int32(name, value, default): Adds an int32 field to the PDU.
+        int64(name, value, default): Adds an int64 field to the PDU.
+        uint64(name, value, default): Adds a uint64 field to the PDU.
+        filler(count): Adds a filler field to the PDU.
+        padding(value): Adds padding to the PDU.
+        fixed_string(name, length, default): Adds a fixed-length string field to the PDU.
+        length_prefixed_string(name, default): Adds a length-prefixed string field to the PDU.
+        variable_length_array(name, element_type, default): Adds a variable-length array field to the PDU.
+        nested_pdu(name, pdu): Adds a nested PDU field to the PDU.
+        compute_crc(data): Computes the CRC32 checksum of the given data.
+        encode(data, compress): Encodes the data into binary format.
+        decode(data, decompress): Decodes the binary data back into the original structure.
```

Comment:
While having this in docs is nice, maintaining this to be consistent with the actual methods implemented may be tricky (as they are in different parts of the code and easy to forget to update). Would it be better to just provide a few examples here and add doc-strings for each method so that those are easier to keep updated?

Reply:
Agreed and updated the documentation as suggested.